src/main/java/io/synadia/flink/source/NatsJetStreamSource.java

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2023 Synadia Communications Inc. All Rights Reserved.
+// See LICENSE and NOTICE file for details.
+
+package io.synadia.flink.source;
+
+import io.synadia.flink.Utils;
+import io.synadia.flink.common.ConnectionFactory;
+import io.synadia.flink.source.enumerator.NatsSourceEnumerator;
+import io.synadia.flink.source.split.NatsSubjectCheckpointSerializer;
+import io.synadia.flink.source.split.NatsSubjectSplit;
+import io.synadia.flink.source.split.NatsSubjectSplitSerializer;
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.connector.source.*;
+import org.apache.flink.api.java.typeutils.ResultTypeQueryable;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class NatsJetStreamSource<OutputT> implements Source<OutputT, NatsSubjectSplit, Collection<NatsSubjectSplit>>, ResultTypeQueryable<OutputT> {
+
+    private final ConnectionFactory connectionFactory;
+    private final String natsSubject;
+    private final DeserializationSchema<OutputT> deserializationSchema;
+    private final NatsConsumeOptions config;
+    private static final Logger LOG = LoggerFactory.getLogger(NatsJetStreamSource.class);
+    private final String id;
+    private final Boundedness mode;
+
+    // Package-private constructor to ensure usage of the Builder for object creation
+    NatsJetStreamSource(DeserializationSchema<OutputT> deserializationSchema, ConnectionFactory connectionFactory, String natsSubject, NatsConsumeOptions config, Boundedness mode) {
+        id = Utils.generateId();
+        this.deserializationSchema = deserializationSchema;
+        this.connectionFactory = connectionFactory;
+        this.natsSubject = natsSubject;
+        this.config = config;
+        this.mode = mode;
+    }
+
+    @Override
+    public TypeInformation<OutputT> getProducedType() {
+        return this.deserializationSchema.getProducedType();
+    }
+
+    @Override
+    public Boundedness getBoundedness() {
+        return this.mode;
+    }
+
+    @Override
+    public SplitEnumerator<NatsSubjectSplit, Collection<NatsSubjectSplit>> createEnumerator(
```

Comment:
If I understand correctly, a split will be a NATS subject. I'm curious about whether Flink will spawn multiple source reader in parallel for a single split?

I'm asking because with Kafka, a split is a partition, and apparently only one consumer will read that partition at a time. NATS is different that multiple consumers can read from the same subject. If I need multiple readers to read a very big subject, how Flink will handle that?

Reply:
Yes the subject is a split. Should only be one reader per split (per subject)

Reply:
If one has only one huge subject, one reader might not be enough to handle the rate. Usually we can spawn multiple nats client to read from the same subject but with this pattern there can only be one I believe.

Reply:
I think we could try to figure out how to use multiple instances of the same pull consumer (works the same as push delivery group). So when that option is configured we would have to know the number of consumer instances and build the splits accordingly.

Reply:
I agree that we should build support for multiple consumers reading from one subject. I will work on it. 
However, for this PR, would prefer to limit it to one subject one consumer. 
We can create an issue in this project as requirement for record keeping.

Reply:
@vu-hoang-phan-form3 @scottf Why can't we have a basic NATS Jetstream Flink connector that reads from the stream with one consumer in this base PR? Multiple consumers for the stream(one for each subject) need new subject discovery, we should take both as improvements together.

Reply:
The intention is to have both. It will have to be some sort of configuration. This needs to be done for the non jetstream as well.

Reply:
Hello @scottf, just to clear if I understand it right, you are suggesting that we should have :
1. one consumer for every subject under a regex subject like (ORDER.*) ?
2.one consumer for every subject from list of subjects that user will provide ?
3.multiple copies of one consumer to increase the read throughput, if subject is big ?

If I have understood it correctly you are suggesting to have multiple copies of one consumer, right ?

Reply:
The original implementation made one split per subject filter. I'm suggesting that we need some sort of configuration hint that tell us to make N splits for a single subject filter. I suppose that could be accomplished by allowing the subject filter multiple times in the input list.
Also, I would not characterize a subject like "order.*" as a regex, it's a wildcard. As presented, it is simply a single subject filter to a consumer.

Reply:
Yes order.* is one subject and thats how I see it too. Now I have got more clarity on this multiple consumer part which I was confusing myself with some other usecase.
Let me see it once again how we can fit creating multiple instance of same consumer for one subject.

Reply:
What if we don't need to have split at all? NATS supports multiple readers out-of-the-box already. So a NATS source just need to create multiple readers with the same config (same stream, same subject) and they will work in parallel without further config.

Reply:
Hi @vu-hoang-phan-form3  It does make sense to ignore split if we do not want to use checkpointing, which is the current implementation, and we are not making use of the Split. 
But to implement the checkpointing we need Split, it holds the data which is fetched from source and same data needs to be provided back to Flink during snapshot.
I have changed the code to implement checkpoint here 
https://github.com/souravagrawal/flink-connector-nats/commit/9c7c54546c8edff0e8a3482a5d3787cd3be6512c

Please go through with it once, I plan to merge it in somrat's fork soon.